Add xtras::SendNext trait to schedule prioritised messages

klochowicz opened the pull request:
`xtras::SendAsyncSafe` by default does not take priority into account, so the
message dispatched by an actor will get scheduled on the back of the queue.

send_next() will dispatch messages with fixed priority
INTERNAL_MSG_PRIORITY (currently set to 100). This trait should be preferred
over send_async_safe() in cases where state changes as a result of the message
    handler.

Examples where send_next() was used in this changeset:
- Subscribers want to know about changes in the Endpoint before they handle a
    new external requests (because e.g. a connection got dropped in the meantime)
- Endpoint wants to process newly created/dropped connection as soon as it can,
    instead of scheduling it on the queue.

bonomat commented on xtras/src/send_next.rs:
```diff
@@ -0,0 +1,51 @@
+pub trait SendNext<M: Send + 'static, R> {
+    /// Dispatch a prioritised message to an actor without waiting for it to handle it.
+    // If the actor is not connected, it logs a warning.
+    async fn send_next(&self, msg: M);
```
I don't really want to participate in a naming discussion but since you asked for it :D 
How about calling it `send_now`?

da-kami replied:
I would probably call it `send_prioritized` - in case we opt for letting the caller decide the priority it could be `send_with_priority`. Don't have a strong opinion about this.

luckysori replied:
None of these names (including the original) tell me that we won't wait for the result. Which I think `send_async_safe` tried to do with the word `async`.

klochowicz replied:
based on the feedback above, I'll go with `send_async_next()` (because this encapsulates that we won't wait for result).

Restioson replied:
That sounds like a good name. I was going to write my own comment about this but @luckysori expressed it perfectly

bonomat commented on xtras/src/send_next.rs:
```diff
@@ -0,0 +1,51 @@
+    M: Send + 'static,
+{
+    #[instrument(skip_all)]
+    async fn send_next(&self, msg: M) {
```
Since it's meant to be used as a prioritized message, why don't you return an error and let the caller decide?

klochowicz replied:
IMHO there is no error strategy / decision that can really be made in such context - we never did anything other than logging it (if we remembered to do so :/ ).
also, in this case, the API with an error could falsely suggest that the message got handled if it returns `Ok()`, but it's not the case (it just means that the actor got it delivered on the inbox.

luckysori commented on xtras/src/send_next.rs:
```diff
@@ -0,0 +1,51 @@
+use xtra::message_channel::MessageChannel;
+use xtra::refcount::RefCounter;
+
+/// Arbitrarily hard number allowing flexibility of prioritising messages in actors
```
❓ Did you mean high?

```suggestion
/// Arbitrarily high number allowing flexibility of prioritising messages in actors
```

da-kami reviewed (approved):
Is there a (simple) way to test that this is working as expected? 
The actor tests pass, but they passed before as well. I feel it would be great to crate a load test that shows that this actually helps with avoiding congestion. (Not part of the scope of this PR, but wanted to mention it)

luckysori reviewed (commented):
I'm not against this change, but I think I would be more in favour if we (in order of importance):

1. Introduced each change in priority by commit, explaining why.
2. Used an internal constant per module. A single constant for all internal messages ever is almost certainly fine too. 

The second point means I'm not over the moon about the trait. I like that it's very descriptive if we say something like `send_now`, but the name quickly loses its meaning as soon as we decide that any message should have a higher priority and we have to set it manually. I would inline the trait, but it's not a deal breaker.

luckysori reviewed (approved):
Approved because I think we're on the same page.

klochowicz committed:
Add xtras::SendNext trait to schedule prioritised messages

`xtras::SendAsyncSafe` by default does not take priority into account, so the
message dispatched by an actor will get scheduled on the back of the queue.

send_next() will dispatch messages with fixed priority
INTERNAL_MSG_PRIORITY (currently set to 100). This trait should be preferred
over send_async_safe() in cases where state changes as a result of the message
    handler.

klochowicz committed:
Use send_async_next when we dispatch an internal task to finish processing

In many places where we use `send_async_safe` we want to dispatch a follow-up
task that is still relevant to the current request we're handling, but we do
not want to block the message handler. It is convenient to schedule such tasks
as messages, but in order to guarantee they would not be subjected to
backpressure or queued up behind subsequent requests (as we want to action them
when their context is still relevant), we need to be able to schedule them with
higher priority than regular messages arriving on our inbox.

Concrete examples from this changeset:
- Endpoint wants to process newly created/dropped connection as soon as it can,
instead of scheduling it on the queue.
- updating metrics after a handler

klochowicz committed:
Use send_async_next when we dispatch messages to subscribers

Subscribers need to know about changes in the Endpoint before they handle
new external requests (because e.g. a connection got dropped in the meantime,
and we might not need to initiate protocol with some takers)
